samples/subsys/usb/uac2_explicit_feedback/src/main.c

```diff
@@ -0,0 +1,683 @@
+
+static void reset_feedback_counter(struct usb_i2s_ctx *ctx)
+{
+	/* Reset feedback to nominal value */
+	ctx->fb_value = SAMPLES_PER_SOF << (FEEDBACK_K + FEEDBACK_FS_SHIFT);
+	if (IS_ENABLED(CONFIG_USE_I2S_LRCLK_EDGES_COUNTER)) {
+		ctx->fb_counter = 0;
+		ctx->fb_periods = 0;
+	} else {
+		ctx->integrator = 0;
+	}
+}
+
+static void uac2_terminal_update_cb(const struct device *dev, uint8_t terminal,
+				    bool enabled, bool microframes,
+				    void *user_data)
+{
+	struct usb_i2s_ctx *ctx = user_data;
+
+	/* This sample has only one terminal therefore the callback can simply
+	 * ignore the terminal variable.
+	 */
+	__ASSERT_NO_MSG(terminal == HEADPHONES_OUT_TERMINAL_ID);
+	/* This sample is for Full-Speed only devices. */
+	__ASSERT_NO_MSG(microframes == false);
+
+	ctx->terminal_enabled = enabled;
+	if (ctx->i2s_started && !enabled) {
+		i2s_trigger(ctx->i2s_dev, I2S_DIR_TX, I2S_TRIGGER_DROP);
+		ctx->i2s_started = false;
+		ctx->i2s_blocks_written = 0;
+		reset_feedback_counter(ctx);
+	}
+}
+
+static void *uac2_get_recv_buf(const struct device *dev, uint8_t terminal,
+			       uint16_t size, void *user_data)
+{
+	ARG_UNUSED(dev);
+	struct usb_i2s_ctx *ctx = user_data;
+	void *buf = NULL;
+	int ret;
+
+	if (terminal == HEADPHONES_OUT_TERMINAL_ID) {
+		__ASSERT_NO_MSG(size <= MAX_BLOCK_SIZE);
+
+		if (!ctx->terminal_enabled) {
+			LOG_ERR("Buffer request on disabled terminal");
+			return NULL;
+		}
+
+		ret = k_mem_slab_alloc(&i2s_tx_slab, &buf, K_NO_WAIT);
+		if (ret != 0) {
+			buf = NULL;
+		}
+	}
+
+	return buf;
+}
+
+static void uac2_data_recv_cb(const struct device *dev, uint8_t terminal,
+			      void *buf, uint16_t size, void *user_data)
+{
+	struct usb_i2s_ctx *ctx = user_data;
+	int ret;
+
+	if (!ctx->terminal_enabled) {
+		k_mem_slab_free(&i2s_tx_slab, buf);
+		return;
+	}
+
+	if (!size) {
+		/* Zero fill to keep I2S going. If this is transient error, then
+		 * this is probably best we can do. Otherwise, host will likely
+		 * either disable terminal (or the cable will be disconnected)
+		 * which will stop I2S.
+		 */
+		size = BLOCK_SIZE;
+		memset(buf, 0, size);
+		sys_cache_data_flush_range(buf, size);
+	}
+
+	LOG_DBG("Received %d data to input terminal %d", size, terminal);
+
+	ret = i2s_write(ctx->i2s_dev, buf, size);
+	if (ret < 0) {
+		ctx->i2s_started = false;
+		ctx->i2s_blocks_written = 0;
+		reset_feedback_counter(ctx);
+
+		/* Most likely underrun occurred, prepare I2S restart */
+		i2s_trigger(ctx->i2s_dev, I2S_DIR_TX, I2S_TRIGGER_PREPARE);
+
+		ret = i2s_write(ctx->i2s_dev, buf, size);
+		if (ret < 0) {
+			/* Drop data block, will try again on next frame */
+			k_mem_slab_free(&i2s_tx_slab, buf);
+		}
+	}
+
+	if (ret == 0) {
+		ctx->i2s_blocks_written++;
+	}
+}
+
+static void uac2_buf_release_cb(const struct device *dev, uint8_t terminal,
+				void *buf, void *user_data)
+{
+	/* This sample does not send audio data so this won't be called */
+}
+
+#if IS_ENABLED(CONFIG_SOC_COMPATIBLE_NRF5340_CPUAPP)
```

Comment:
I agree. I just didn't have an idea how to cleanly separate it, i.e. how the "interface" between this target specific code and rest of application should look like. I can try to think of something but not sure if I'll come up with anything better than just inline ifdef.

Reply:
can it be solved without using anything nrfx specific?

Reply:
I don't think so. The only non-specific solution would be to gather tick timestamps in both I2S and USB interrupts. The problematic part is to expose such timestamp information to application and to actually capture the tick timestamp inside the intterupt handler. This would still most likely need additional logic to filter out for software scheduling overhead (to function properly this needs sub-sample precision, here the PI controller gets 1/32 sample estimate from nRF specific code).

Reply:
Of course, I would like to be proven wrong. But to prove me wrong someone would have to actually come up with actual implementation that works. I agree that is is not ideal to have target-specific code in samples, but here for the time being it seems to be the only sensible approach. The sample can be compiled for other targets (with a compile time warning) and it will operate but will suffer underruns/overruns because the lack of proper feedback (which must be computed at run-time, because it essentially is ratio between actual I2S master clock and USB host SOF clock, i.e. nominally both are the same, but the actual deviations from nominal value is the key here).

For the USB Audio 2 class implementation however, the explicit feedback value is just something that application **must** provide (unless it either uses implicit feedback in which case it has to send audio samples at appropriate rate to hsot; or unless it claims that the clock is `sof-synchronized` in which case there'll always be nominal number of samples). The sample is in my opinion mandatory for such non-obvious functionality. While it would be best to have it target independent, it is always helpful to have some reference implementation that is known to work.

Reply:
I don't think we would want a sample living among the generic samples to be specific to Nordic? I had a quick look and I would say that so far we seem to have done a pretty good job at avoiding that (not sure about [this guy](https://github.com/zephyrproject-rtos/zephyr/blob/9948c29885860b9800257f5a1fd466eb046b426e/samples/subsys/ipc/ipc_service/icmsg/src/main.c#L133), but that seems to be pretty much it) .
Could this be moved to `boards/nrf`? Is the "common" (and I understand maybe degraded?) version of this sample still something thart could live in an agnostic way in the locaiton currently proposed?

Reply:
@kartben  We could remove the sample altogether then because without target specific code it does not make sense to have it. We could then just have class implementation alone without any sample. Would doing so be any better?

Reply:
@kartben Not sure if you noticed but the sample is not really specific to Nordic as it is now. It does have the `#else` with warning and empty implementation of the necessary feedback functionality. This shows how to do the feedback, with two different methods, on Nordic which is IMHO better than removing the Nordic specific part and leaving all targets essentially not doing that this sample is supposed to show. Because the targeet specific code is also extremely application (i.e. sample) specific it doesn't make slightest sense to put it in `boards/nrf`.